Collect and display execution metadata for ES|QL cross cluster searches

quux00 opened the pull request:
Enhance ES|QL responses to include information about `took` time (search latency), shards, and clusters against which the query was executed.

The goal of this PR is to begin to provide parity between the metadata displayed for cross-cluster searches in _search and ES|QL.

This PR adds the following features:
- add overall `took` time to all ES|QL query responses. And to emphasize: "all" here means: async search, sync search, local-only and cross-cluster searches, so it goes beyond just CCS.
- add `_clusters` metadata to the final response for cross-cluster searches, for both async and sync search (see example below)
- tracking/reporting counts of skipped shards from the can_match (SearchShards API) phase of ES|QL processing
- marking clusters as skipped if they cannot be connected to (during the field-caps phase of processing)

Out of scope for this PR:
- honoring the `skip_unavailable` cluster setting
- showing `_clusters` metadata in the async response **while** the search is still running
- showing any shard failure messages (since any shard search failures in ES|QL are automatically fatal and _cluster/details is not shown in 4xx/5xx error responses). Note that this also means that the `failed` shard count is always 0 in ES|QL `_clusters` section.

Things changed with respect to behavior in `_search`:
- the `timed_out` field in `_clusters/details/mycluster` was removed in the ESQL response, since ESQL does not support timeouts. It could be added back later if/when ESQL supports timeouts.
- the `failures` array in `_clusters/details/mycluster/_shards` was removed in the ESQL response, since any shard failure causes the whole query to fail.

Example output from ES|QL CCS:

```es
POST /_query
{
  "query": "from blogs,remote2:blo*,remote1:blogs|\nkeep authors.first_name,publish_date|\n limit 5"
}
```

```json
{
  "took": 49,
  "columns": [
    {
      "name": "authors.first_name",
      "type": "text"
    },
    {
      "name": "publish_date",
      "type": "date"
    }
  ],
  "values": [
    [
      "Tammy",
      "2009-11-04T04:08:07.000Z"
    ],
    [
      "Theresa",
      "2019-05-10T21:22:32.000Z"
    ],
    [
      "Jason",
      "2021-11-23T00:57:30.000Z"
    ],
    [
      "Craig",
      "2019-12-14T21:24:29.000Z"
    ],
    [
      "Alexandra",
      "2013-02-15T18:13:24.000Z"
    ]
  ],
  "_clusters": {
    "total": 3,
    "successful": 2,
    "running": 0,
    "skipped": 1,
    "partial": 0,
    "failed": 0,
    "details": {
      "(local)": {
        "status": "successful",
        "indices": "blogs",
        "took": 43,
        "_shards": {
          "total": 13,
          "successful": 13,
          "skipped": 0,
          "failed": 0
        }
      },
      "remote2": {
        "status": "skipped",  // remote2 was offline when this query was run
        "indices": "remote2:blo*",
        "took": 0,
        "_shards": {
          "total": 0,
          "successful": 0,
          "skipped": 0,
          "failed": 0
        }
      },
      "remote1": {
        "status": "successful",
        "indices": "remote1:blogs",
        "took": 47,
        "_shards": {
          "total": 13,
          "successful": 13,
          "skipped": 0,
          "failed": 0
        }
      }
    }
  }
}
```

Fixes https://github.com/elastic/elasticsearch/issues/112402 and https://github.com/elastic/elasticsearch/issues/110935

quux00 committed:
Collect and display execution metadata for ES|QL cross cluster searches

Added EsqlExecutionInfo class and added it to the EsqlQueryResponse,
but not yet in the XContent output of EsqlQueryResponse.

Also added to TransportEsqlQueryAction, but not yet passed down to the execution layers.

quux00 committed:
Slight improvements to EsqlExecutionInfo

quux00 committed:
Removed changes to EsqlQueryResponse, spending too long getting the EsqlQueryResponseTests to pass and probably don't want it in the response for now anyway

quux00 committed:
Starting threading EsqlExecutionInfo into PlanExecutor and EsqlSesssion, but not used yet to record metadata

quux00 committed:
Have the initial swap-in of cluster info into EsqlExecutionInfo in EsqlSession.preAnalyzeIndices. No updates after that added yet.

Passing in Function<String, Map<String, OriginalIndices>> rather than RemoteClusterService to PlanExecutor and EsqlSession.
Makes it easier (or possible) to get the existing unit tests working again and less overall coupling.

quux00 committed:
Added EsqlExecutionInfo to IndexResolver. Enrich pathway passes in null for EsqlExecutionInfo.

The EsqlExecutionInfo Cluster statuses are update in IndexResolver.updateExecutionInfoWithFieldCapsResults.
Test for that method added.

quux00 commented:
buildkite test this

quux00 committed:
ComputeListener updated to the version that has proper remote/local split and new static constructors.

However, the new acquire fns in ComputeListener are not yet used in ComputeService.
ComputeResponse updated to have info (shard accounting and took time) need for adding to EsqlExecutionInfo,
but again not being used/added yet.
Added to queryStartTimeMillis to esql.session.Configuration

quux00 committed:
Added new tests to ComputeListenerTests

And adjusted acquireComputeForDataNodes to take long for startTime rather than whole Configuration object

quux00 committed:
Added ExecutionInfo to Result obj (used in ComputeService/EsqlSession)

quux00 committed:
update ExecutionInfo with shard counts in ComputeService.lookupDataNodes

Added new DataNodeResult record to be returned from ComputeService.lookupDataNodes

quux00 committed:
Migrated CrossClustersQueryIT to new setup format, but can't add execInfo tests until its added to the EsqlQueryResponse

quux00 committed:
Added CountDown to acquireComputeForDataNodes - that allows SUCCESSFUL status to be set for local cluster when last node finishes

Added EsqlExecutionInfo to EsqlQueryResponse, but only took time is in the XContent
Added new tests and test cases to CrossClustersQueryIT

quux00 committed:
Fixed failing REST and qa tests to account for the new 'took' time in the ESQL response

quux00 committed:
Fixed bug where CountDown in ComputeService can be initialized with 0 count

quux00 committed:
More qa and bwc test fixes based on what failed in latest ci build

quux00 committed:
Next round of qa and bwc test fixes based on what failed in latest ci build

quux00 committed:
Fix failing test in EsqlSecurityIT

quux00 committed:
Added _cluster/details to the EsqlQueryResponse XContent for cross-cluster searches

quux00 committed:
Fixed test failure in esql/ccq/MultiClustersIT

quux00 committed:
Updated end user docs with info about top level took time and _cluster/details section

quux00 committed:
Added EsqlExecutionInfo to equals and hashCode method of EsqlQueryResponse and got EsqlQueryResponseTests passing

quux00 committed:
Removed skip_unavilable=true filter in IndexResolver - all clusters are marked as skipped for now if remote cannot be contacted

quux00 committed:
Moved isRemoteUnavailableException to ExceptionsHelper

quux00 committed:
Added equals and hashCode to EsqlExecutionInfo.Cluster object.

Improved javadoc and fixed a few minor test flaws.
Removed some debugging code.

quux00 committed:
Minor tweak to esql-across-clusters.asciidoc

quux00 committed:
Improvements to esql-across-clusters.asciidoc

quux00 commented:
Question for reviewers:

Do we want `took` time in millis or nanos?

**The argument for using `millis`:**

`took` time will be consistent with `_search` which uses millis, so I've gone with millis, since my goal was to keep the metadata like `_search` as much as possible unless we decide to change some parts. 

This potentially makes it simpler for Kibana as well if they already have code that parses and displays took time.

**The argument for using `nanos`:**

ESQL Driver Profiles use took_nanos. So should we also use nanos for overall search latency? Will using millis feel inconsistent from profiles to end users?

quux00 commented:
Question for reviewers / Product Managers:

**Do we want to have the partial status in the ESQL metadata? Is that possible to have now? Will it ever be possible?**

I've kept all the cluster statuses that we use on the _search side. These are the meanings of those statuses in _search:

successful: All shards searched returned successfully
running: Search is still running on that cluster
partial: At least one shard search failed or the search timed out on the server side (and thus may have searched all index segments), so we are returning partial data.
skipped: No shards were searched because the cluster was offline and marked as skip_unavailable=true
failed: The cluster was offline and marked as skip_unavailable=false. This causes the whole search to fail (return 4xx/5xx)

Do we want to keep these same status meanings in ES|QL?

quux00 commented:
Edge case question for reviewers / Product Managers:

**What should the status of the cluster be in the following scenario?**

The user makes a cross-cluster query where the index expression on a cluster matches no indices:

```
POST /_query
{
  "query": "from blogs,remote2:no_such_index,remote1:blogs|\nkeep authors.first_name,publish_date|\n limit 5"
}
```

Currently, in the code in this PR, I mark this as "SUCCESSFUL" (see example below), but with total shards = 0, indicating that nothing was searched. Is that the behavior we want? Or should we mark it as "SKIPPED"?


<details><summary>(toggle) **ES|QL response when cluster specified with index expression that matching no indices**</summary>

```json


{
  "took": 25,
  "columns": [
    {
      "name": "authors.first_name",
      "type": "text"
    },
    {
      "name": "publish_date",
      "type": "date"
    }
  ],
  "values": [
    [
      "Theresa",
      "2020-09-14T23:13:55.000Z"
    ],
  ...
  ],
  "_clusters": {
    "total": 3,
    "successful": 2,
    "running": 0,
    "skipped": 1,
    "partial": 0,
    "failed": 0,
    "details": {
      "(local)": {
        "status": "successful",
        "indices": "blogs",
        "took": 24,
        "_shards": {
          "total": 13,
          "successful": 13,
          "skipped": 0,
          "failed": 0
        }
      },
      "remote2": {
        "status": "successful",
        "indices": "remote2:blogs",
        "took": 24,
        "_shards": {
          "total": 4,
          "successful": 4,
          "skipped": 0,
          "failed": 0
        }
      },
      "remote1": {
        "status": "skipped",
        "indices": "remote1:blogs",
        "took": 0,
        "_shards": {
          "total": 0,
          "successful": 0,
          "skipped": 0,
          "failed": 0
        }
      }
    }
  }
}
```

</details>


#### Comparison to behavior in `_search`

In `_search`, the handling of this differs depending on whether the user used a wildcard or not. With a wildcard that matches nothing you get status=successful, _shards.total=0:

```
        "remote1": {
          "status": "successful",
          "indices": "x*",
          "took": 0,
          "timed_out": false,
          "_shards": {
            "total": 0,
            "successful": 0,
            "skipped": 0,
            "failed": 0
          }
        }
```

With no wildcard, you get status=skipped and a no_such_index failure entry:

```
        "remote1": {
          "status": "skipped",
          "indices": "my_index",
          "timed_out": false,
          "failures": [
            {
              "shard": -1,
              "index": null,
              "reason": {
                "type": "index_not_found_exception",
                "reason": "no such index [my_index]",
                "index_uuid": "_na_",
                "resource.type": "index_or_alias",
                "resource.id": "my_index",
                "index": "my_index"
              }
            }
          ]
        }
```

elasticsearchmachine commented:
Hi @quux00, I've created a changelog YAML for you.

quux00 committed:
Update docs/changelog/112595.yaml

elasticsearchmachine commented:
Pinging @elastic/es-analytical-engine (Team:Analytics)

quux00 commented on docs/reference/esql/esql-across-clusters.asciidoc:
```diff
@@ -174,6 +174,184 @@ FROM *:my-index-000001
+
+
+
+
```
@naj-h and @tylerperk - Please review proposed end user docs changes.

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:
```diff
@@ -151,13 +168,129 @@ public IndexResolution mergedMappings(String indexPattern, FieldCapabilitiesResp
+                    (k, v) -> new EsqlExecutionInfo.Cluster.Builder(v).setStatus(EsqlExecutionInfo.Cluster.Status.SKIPPED).build()
+                );
+                // TODO: follow-on PR will set SKIPPED status when skip_unavailable=true and throw an exception when skip_unavailable=false
+            }
```
I was not able to find a good way to write a test against this scenario. I did try to set up a test in CrossClustersQueryIT where I created a third remote cluster and then took all of its data nodes offline, resulting in SKIPPED status, but the test was very flaky, so I removed it. I did test this scenario manually several times.

dnhatn replied:
Could you try testing with MockTransportService#addConnectBehavior?

quux00 commented:
buildkite test this

quux00 committed:
Added questions about took time headers to EsqlResponseListener - possibly need to change that impl

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

astefan reviewed (approved):
LGTM
I left only some small bits.

I especially appreciate the part that accounts for missing indices when looking at the _field_caps response in IndexResolver.

Nice 👍 on the testing as well.
One class that I looked into recently (otherwise I didn't know about it) is `RemoteClusterSecurityEsqlIT` (from xpack -> plugin -> security -> qa -> multi-cluster) where I am not sure this work in this PR should reflect there, as well. I have the impression it should be there, but I am not 100% sure.

Might want to wait also on @dnhatn or @nik9000 to have a look, as it's touching mainly the distributed part of ES|QL (my main expertize is around the planner and the IndexResolver).

nik9000 commented:
> **Do we want to have the partial status in the ESQL metadata? Is that possible to have now? Will it ever be possible?**

Right now we fail if any shard fails, so, at least now, we can't have PARTIAL as you defined it. Will it ever be possible? I dunno. Planning isn't my strong suit.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlResponseListener.java:
```diff
@@ -154,6 +154,10 @@ private RestResponse buildResponse(EsqlQueryResponse esqlResponse) throws IOExce
+            // TODO: question for reviewers - should we instead grab took time from the EsqlQueryResponse now and remove this StopWatch?
+            // TODO: that would solve the issue https://github.com/elastic/elasticsearch/issues/110935
+            // TimeValue tookMillis = esqlResponse.getExecutionInfo().getOverallTook();
             restResponse.addHeader(HEADER_NAME_TOOK_NANOS, Long.toString(tookNanos));
```
Yes. I think we should.

quux00 replied:
Thanks Nik. I think this means then that I will need to change the code in this PR to track `took` time in nanos, rather than millis? Or do we want to remove/deprecate this header and put in one just called `took`? (I have a similar question out to Product, but would like to hear your views.)

quux00 replied:
I've updated the internal to track took time by nanos now. However, I left the XContent output in millis, as that is how _search does it. Question is out to @naj-h whether to leave this as millis to change to `took_nanos` in the response object. (The `Took-nanos` header is still present and done in nanos.)

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/enrich/EnrichPolicyResolver.java:
```diff
@@ -363,7 +363,7 @@ public void messageReceived(LookupRequest request, TransportChannel channel, Tas
                     try (ThreadContext.StoredContext ignored = threadContext.stashWithOrigin(ClientHelper.ENRICH_ORIGIN)) {
                         String indexName = EnrichPolicy.getBaseName(policyName);
-                        indexResolver.resolveAsMergedMapping(indexName, IndexResolver.ALL_FIELDS, refs.acquire(indexResult -> {
+                        indexResolver.resolveAsMergedMapping(indexName, IndexResolver.ALL_FIELDS, null, refs.acquire(indexResult -> {
```
Nik: Check if this is important.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeResponse.java:
```diff
@@ -22,8 +23,31 @@
+        this(profiles, null, null, null, null, null);
+    }
+
+    ComputeResponse(
```
I'm curious to see how this all works with INLINESTATS which runs a two phased execution. We're able to preserve the profiles for them with some work but this is outside of all that.

quux00 replied:
As discussed in another forum, I'm not able to test this currently since INLINESTATS itself is not yet cross-cluster enabled, as indicated here: https://github.com/elastic/elasticsearch/issues/110923. We will not block this ticket on that issue and instead handle it as part of the INLINESTATS work.

nik9000 replied:
:+1:

INLINESTATS is too busted at the moment to test this with properly

nik9000 replied:
Sorry for the distraction.

nik9000 reviewed (commented):
I don't think I know enough to approve this, at least not without another read. I've left a few comments and will have another look soon. If @dnhatn can look and approve, great. Otherwise i'll look more soon.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/TransportEsqlQueryAction.java:
```diff
@@ -245,7 +273,8 @@ public EsqlQueryResponse initialResponse(EsqlQueryTask task) {
             true, // is_running
-            true // isAsync
+            true, // isAsync
+            null  // MP TODO: is this what we want here? test this
```
Thanks. Removed them. Fixed in next push.

quux00 commented on x-pack/plugin/esql/qa/server/multi-clusters/src/javaRestTest/java/org/elasticsearch/xpack/esql/ccq/MultiClustersIT.java:
```diff
@@ -145,13 +149,40 @@ public void testCount() throws Exception {
+            assertTrue(result.containsKey("took"));
+            assertTrue(result.containsKey("columns"));
+            assertTrue(result.containsKey("values"));
+            assertTrue(result.containsKey("_clusters"));
```
Thanks. I removed (in the next push) the `extraOk()` call and used mapMatcher more. That allowed removing the 5 lines of code here that is repeated in several places.

quux00 commented on x-pack/plugin/esql/qa/server/single-node/src/javaRestTest/java/org/elasticsearch/xpack/esql/qa/single_node/RestEsqlIT.java:
```diff
@@ -333,14 +340,19 @@ public void testInlineStatsProfile() throws IOException {
+
         Map<String, Object> result = runEsql(builder);
+        MapMatcher mapMatcher = matchesMap();
+        if (result.get("took") != null) {
+            mapMatcher = mapMatcher.entry("took", ((Integer) result.get("took")).intValue());
+        }
```
Thanks. I've removed the null checks. There was only place where it was needed in another test, so now I just add `took` as another entry like normal. Will be in next push.

quux00 committed:
PR feedback with focus on end user docs fixes, removing some out-of-date code comments and minor improvements to tests (qa and IT)

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/Configuration.java:
```diff
@@ -76,6 +78,7 @@ public Configuration(
         this.profile = profile;
         this.tables = tables;
         assert tables != null;
+        this.queryStartTimeMillis = queryStartTimeMillis;
```
Hmm, not sure. If you feel that is safe and future proof. Right now this class (in non-test code) is only created in one place (TransportEsqlQueryAction.innerExecute), so your suggestion would certainly work, but burying it in Configuration feels like a hidden piece of state that should be clear to the Action doing it. On the _search side, for example, there is an explicit [SearchTimeProvider](https://github.com/elastic/elasticsearch/blob/main/server/src/main/java/org/elasticsearch/action/search/TransportSearchAction.java#L294) that gets passed around to make it explicit.

So my vote is that I prefer it to be created in the Action class and passed in. That makes it clear that Configuration is used for tracking time and could allow testing it be easier if we need to pass in stubbed in time values.

Side note: the Configuration class already has an `absoluteStartedTimeInMillis` that returns `System.currentTimeMillis()` each time it is called, which seems like incorrect/unexpected behavior to me. That's a case where I think the Configuration class should not be creating that time (since it doesn't retain it).

astefan replied:
> Side note: the Configuration class already has an absoluteStartedTimeInMillis that returns System.currentTimeMillis() each time it is called, which seems like incorrect/unexpected behavior to me. That's a case where I think the Configuration class should not be creating that time (since it doesn't retain it).

Indeed, that doesn't look right.

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeResponse.java:
```diff
@@ -37,6 +61,24 @@ final class ComputeResponse extends TransportResponse {
         }
+        if (in.getTransportVersion().onOrAfter(TransportVersions.ESQL_CCS_COMPUTE_RESPONSE)) {
+            Long took = in.readOptionalLong();
+            if (took == null) {
+                this.took = null;
+            } else {
+                this.took = new TimeValue(took);
+            }
```
Thanks. Fixed in next push.

quux00 commented on test/external-modules/esql-heap-attack/src/javaRestTest/java/org/elasticsearch/xpack/esql/heap_attack/HeapAttackIT.java:
```diff
@@ -287,7 +308,11 @@ public void testManyConcat() throws IOException {
             columns = columns.item(matchesMap().entry("name", "str" + s).entry("type", "keyword"));
         }
-        assertMap(map, matchesMap().entry("columns", columns).entry("values", any(List.class)));
+        MapMatcher mapMatcher = matchesMap();
```
I was able to remove all this special handling since there is only one place that should expect a null took time in the tests.

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlExecutionInfo.java:
```diff
@@ -0,0 +1,537 @@
+     * For testing use with fromXContent parsing only
+     * @param clusterInfo
+     */
+    public EsqlExecutionInfo(ConcurrentMap<String, Cluster> clusterInfo) {
```
Good idea. Changed in next push.

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlExecutionInfo.java:
```diff
@@ -0,0 +1,537 @@
+        }
+    }
+
+    public void setOverallTookTime(TimeValue took) {
```
This is mimicking a class in core (SearchResponse.Clusters and SearchResponse.Cluster), but I've modified this method to lose the get/set (next push).

quux00 committed:
Additional PR feedback changes - test adjustments, remove 'set' and 'get' prefixes from some methdos in EsqlExecutionInfo

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/action/EsqlExecutionInfo.java:
```diff
@@ -0,0 +1,537 @@
+            this.skippedShards = in.readOptionalInt();
+            this.failedShards = in.readOptionalInt();
+            Long took = in.readOptionalLong();
+            if (took == null) {
+                this.took = null;
+            } else {
+                this.took = new TimeValue(took);
+            }
```
Thanks Nik. Changed to using read/writeOptionalTimeValue in next push. Good to know that exists!

quux00 committed:
Now tracking took in nanos, not millis (but XContent still displays in millis, until a decision is made to change it)

Also reading/writing TimeValue with read/writeOptionalTimeValue.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 committed:
Changed ComputeResponse to de/serialize with read/writeOptionalTimeValue

quux00 committed:
EsqlResponseListener now preferentially uses the took time in the EsqlExecutionInfo; falls back to internal Stopwatch

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 committed:
Modified esql-across-clusters to run the new queries I added; but JSON responses are not tested - not clear how to do multiline replacements in this test framework

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 commented on docs/reference/esql/esql-across-clusters.asciidoc:
```diff
@@ -174,6 +174,184 @@ FROM *:my-index-000001
+Here we show an example using the async search endpoint. {ccs-cap} metadata is also present in the synchronous
+search endpoint.
+
+[source,esql]
```
OK, I have changed the two queries I added to actually be executed, but (see comment below) I am unable to find a way to test the response JSON, so I have left that as "TEST[skip...]".

quux00 commented on docs/reference/esql/esql-across-clusters.asciidoc:
```diff
@@ -174,6 +174,184 @@ FROM *:my-index-000001
+  }
+}
+----
+// TEST[skip: cross-cluster testing env not set up]
```
I spent several hours trying but unless you know of a trick to do clever multi-line matching I don't see how this is possible. Among the things I tried was adding "m" to the end of the matcher to indicate multi-line matching (as in Perl matching), but that doesn't work. Mostly I just get failed runs with no information as to what is wrong. 

Plus I'm not really sure it's worth it? The whole point of this section is to show the `_clusters/details` section so testing against a non-CCS set up doesn't seem useful. 

We probably need another ticket to enable the multi-cluster testing setup that search-across-clusters.asciidoc uses, as that was not set up for this test.

nik9000 replied:
:+1:

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

dnhatn reviewed (commented):
I've left some comments. Thanks Michael.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:
```diff
@@ -151,13 +165,110 @@ public IndexResolution mergedMappings(String indexPattern, FieldCapabilitiesResp
+        if (failures != null) {
+            Set<String> unavailableClusters = new HashSet<>();
+            for (FieldCapabilitiesFailure failure : failures) {
+                if (ExceptionsHelper.isRemoteUnavailableException(failure.getException())) {
```
If we (ES|QL) fully fork FieldCaps that would very doable. I'm leery of changing the core field-caps since that is used in other contexts and this PR is large enough that I wouldn't want to take it on in this ticket. I'll put a code comment here about that and create a GH issue to add that feature to either field-caps or EsqlResolveFieldsAction.

quux00 replied:
Issue created: https://github.com/elastic/elasticsearch/issues/113394

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:
```diff
@@ -151,13 +165,110 @@ public IndexResolution mergedMappings(String indexPattern, FieldCapabilitiesResp
+                if (v == null) {
+                    return new StringBuilder().append(indexExpression);
+                } else {
+                    return v.append(',').append(indexExpression);
```
Good point about it possibly being overly long and expensive. We don't need to, but since it was easy to add here, I tried to do it. But _search `_clusters/details` does not show the resolved indices - just the requested index expression from the user (so keeping the wildcards if any present), so I can remove this part and just keep the original index expression. That would make it consistent with _search.

quux00 committed:
Removed code that lists fully resolved indices in the _clusters/details (PR feedback) and updated tests

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 committed:
Code cleanup - remove commented out code in IndexResolverTests

quux00 committed:
PR feedback: Moved logic for unavailable/missing clusters to EsqlSession

EsqlExecutionInfo is no longer passed into IndexResolver. Instead it returns the info
needed in IndexResolution so that ExecInfo.Cluster info can be updated in EsqlSession.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeListener.java:
```diff
@@ -84,6 +142,70 @@ ActionListener<ComputeResponse> acquireCompute() {
+            esqlExecutionInfo.swapCluster(clusterAlias, (k, v) -> {
+                EsqlExecutionInfo.Cluster.Builder builder = new EsqlExecutionInfo.Cluster.Builder(v);
+                if (v.getTook() == null || v.getTook().nanos() < tookOnDataNode.nanos()) {
+                    builder.setTook(tookOnDataNode);
```
Thanks for flagging this. Are you referring to processing done on the coordinator after the data nodes finished, such as a `stats` step? If yes, that would be handled in the `acquireCompute()` call that is passed into `runCompute` - is that correct?

dnhatn replied:
I meant this compute: https://github.com/elastic/elasticsearch/blob/6d076dfa17438c02fdd4e912549ff26e57334c72/x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java#L809-L814

quux00 replied:
Fixed.

quux00 committed:
PR feedback: I removed acquireCCSCompute and acquireComputeForDatanodes in ComputeListener.

This however requires adding additional state to ComputeListener to know what "context" each compute
listener is running so it knows what actions to take.

mark-vieira commented on .idea/inspectionProfiles/Project_Default.xml:
```diff
@@ -1,6 +1,11 @@
 <component name="InspectionProjectProfileManager">
   <profile version="1.0">
     <option name="myName" value="Project Default" />
+    <inspection_tool class="DuplicatedCode" enabled="true" level="WEAK WARNING" enabled_by_default="true">
```
Was committing this intentional?

quux00 replied:
No, I'll remove it.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2-26-Sep-SQUASHED

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -88,6 +93,7 @@ public class EsqlSession {
     private final Mapper mapper;
     private final PhysicalPlanOptimizer physicalPlanOptimizer;
     private final PlanningMetrics planningMetrics;
+    private final Function<String, Map<String, OriginalIndices>> resolveClusterAndIndicesFunction;
```
OK, I've created `IndicesExpressionResolver` interface and have RemoteClusterService implementing it - in next push.

quux00 replied:
Update: based on discussion with @jbaiera I've renamed this `IndicesExpressionGrouper` since it doesn't actually do index resolution, just grouping.

quux00 committed:
PR feedback: Created new intf IndicesExpressionResolver and have RemoteClusterService implementing it

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:
```diff
@@ -75,16 +80,25 @@ public IndexResolver(Client client) {
+    public void resolveAsMergedMapping(
+        String indexWildcard,
+        Set<String> fieldNames,
+        EsqlExecutionInfo executionInfo,
```
Yes, I have changed that now.

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeListener.java:
```diff
@@ -84,6 +142,70 @@ ActionListener<ComputeResponse> acquireCompute() {
+     * @param queryStartTimeNanos start time (on coordinating cluster) for computing took time (per cluster)
+     * @param countDown counter of number of data nodes to wait for before changing cluster status from RUNNING to SUCCESSFUL
+     */
+    ActionListener<ComputeResponse> acquireComputeForDataNodes(String clusterAlias, long queryStartTimeNanos, CountDown countDown) {
```
I have made a go at doing that in this commit: https://github.com/elastic/elasticsearch/pull/112595/commits/fc53eb797c047ea2f899f61ccd64e0a4c527f2e4

To compensate for having the 4 use cases for this method all mixed into the same method, I had to add more state and various helper methods to allow the caller know what "context" it is in in order to determine what (if any) metadata it should be gathering and updating.

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeListener.java:
```diff
@@ -84,6 +142,70 @@ ActionListener<ComputeResponse> acquireCompute() {
+     * @return Listener that will fill in all metadata from to remote cluster into the
+     *         {@link EsqlExecutionInfo}  for the clusterAlias cluster.
+     */
+    ActionListener<ComputeResponse> acquireCCSCompute(String clusterAlias) {
```
Removed in this commit https://github.com/elastic/elasticsearch/pull/112595/commits/fc53eb797c047ea2f899f61ccd64e0a4c527f2e4

quux00 committed:
checkstyle fix

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 committed:
Moved parseClusterAlias from IndexResolver to RemoteClusterAware and using the new methods in RemoteClusterAware

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeListener.java:
```diff
@@ -40,19 +46,117 @@ final class ComputeListener implements Releasable {
+    private final EsqlExecutionInfo esqlExecutionInfo;
+    private final long queryStartTimeNanos;
+    // clusterAlias indicating where this ComputeListener is running
+    // used by the top level ComputeListener in ComputeService on both local and remote clusters
```
I'd prefer javadoc for this.

quux00 replied:
Fixed.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeListener.java:
```diff
@@ -40,19 +46,117 @@ final class ComputeListener implements Releasable {
+        return new ComputeListener(transportService, task, null, null, -1, delegate);
+    }
+
+    // for use by top level ComputeListener in ComputeService
```
Same here.

quux00 replied:
Fixed.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeResponse.java:
```diff
@@ -11,19 +11,56 @@
+    }
+
+    @Override
+    public String toString() {
```
Could you put this further down? Just not between the ctors.

quux00 replied:
Yeah, sorry. Debugging left over. I'll remove it since I don't need it any more.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeListener.java:
```diff
@@ -71,19 +175,56 @@ ActionListener<Void> acquireAvoid() {
+     *                            into the {@link EsqlExecutionInfo} Cluster objects.
      */
-    ActionListener<ComputeResponse> acquireCompute() {
+    ActionListener<ComputeResponse> acquireCompute(@Nullable String computeClusterAlias) {
```
Could you add an explanation to the javadoc about when this is `null` vs the special local key?

nik9000 replied:
Looks like when it's null we don't gather any metadata at al.

quux00 replied:
Good idea. Added in next push.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/Configuration.java:
```diff
@@ -163,9 +174,17 @@ public String query() {
      * Note: Currently, it returns {@link System#currentTimeMillis()}, but this value will be serialized between nodes.
      */
     public long absoluteStartedTimeInMillis() {
+        // MP TODO: I'm confused - Why is this not a fixed value taken at the start of the query processing?
```
Yikes. I'll bet this is old. It should totally be serialized too. *looks* like a bug to me.

nik9000 replied:
But not as part of this PR.

nik9000 replied:
I opened https://github.com/elastic/elasticsearch/issues/113709

nik9000 replied:
I think we could just use `queryStartTimeMillis` here. but not in this PR because I have no idea if it'd do something unexpected.

quux00 replied:
Yeah, these two fields/methods feel redundant but this PR has so much going on I didn't want to futz with it here since the pre-existing case is unclear to me. Thanks for creating a follow-on ticket.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -242,7 +264,51 @@ private <T> void preAnalyze(LogicalPlan parsed, BiFunction<IndexResolution, Enri
+        for (String c : clustersWithNoMatchingIndices) {
+            executionInfo.swapCluster(
+                c,
+                (k, v) -> new EsqlExecutionInfo.Cluster.Builder(v).setStatus(EsqlExecutionInfo.Cluster.Status.SUCCESSFUL)
```
It'd be nicer if these were `NOT_MATCHING`. Could you get that into this one? I know it's not super important, but it'd save some serialization changes.

quux00 replied:
I'll take that one to Product. I have an open question with product on how to handle this case - this adds a fourth option.

nik9000 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/EsqlSession.java:
```diff
@@ -252,6 +318,16 @@ private void preAnalyzeIndices(LogicalPlan parsed, ActionListener<IndexResolutio
+                String indexExpr = Strings.arrayToCommaDelimitedString(entry.getValue().indices());
+                executionInfo.swapCluster(clusterAlias, (k, v) -> {
+                    assert v == null : "No cluster for " + clusterAlias + " should have been added to ExecutionInfo yet";
+                    return new EsqlExecutionInfo.Cluster(clusterAlias, indexExpr.toString(), executionInfo.isSkipUnavailable(clusterAlias));
```
`indexExpr` is already a string. `toString` isn't doing much here I think.

quux00 replied:
Thanks for flagging. Fixed in next push.

nik9000 reviewed (approved):
I think it's a big improvement. I asked for a few things, but I don't think we need another review.

quux00 committed:
Renamed IndicesExpressionResolver intf to IndicesExpressionGrouper.

Added ExecutionInfo asserts to CrossClustersEnrichIT.

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 committed:
PR feedback: Added javadoc to ComputeListener, removed leftover debugging code, other minor code cleanup.

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeListener.java:
```diff
@@ -41,14 +46,67 @@ final class ComputeListener implements Releasable {
+        EsqlExecutionInfo executionInfo,
+        ActionListener<ComputeResponse> delegate
+    ) {
+        return new ComputeListener(transportService, task, clusterAlias, executionInfo, delegate);
```
I think this request is now out of date/not needed with the new restructuring of the code?

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/session/IndexResolver.java:
```diff
@@ -151,13 +165,110 @@ public IndexResolution mergedMappings(String indexPattern, FieldCapabilitiesResp
+            String indexExpression = ir.getIndexName();
             concreteIndices.put(ir.getIndexName(), ir.getIndexMode());
+
+            String clusterAlias = parseClusterAlias(indexExpression);
```
Fixed.

astefan reviewed (approved):
LGTM
Sorry for the long wait.

quux00 committed:
Fixed bug where SKIPPED status for unavailable clusters from field-caps lookup was being overwritten to SUCCESSFUL status

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

dnhatn commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeListener.java:
```diff
@@ -40,19 +46,132 @@ final class ComputeListener implements Releasable {
+    /**
+     * @return true if the "local" querying/coordinator cluster is being searched in a cross-cluster search
+     */
+    private boolean coordinatingClusterIsSearchedInCCS() {
```
I think we can remove the methods `coordinatingClusterIsSearchedInCCS`, `runningOnRemoteCluster`, and `isCCSListener`, `shouldRecordTookTime` as we should be able to collect execution info similarly to how we handle profiles and warnings. These methods require careful consideration of when and how to pass the `clusterAlias`, which I think increases the risk of errors. That said, it's entirely up to you whether to keep them as they are or remove them.

quux00 replied:
I'm not sure that's true. I find the ComputeListener very hard to reason about, so I could be wrong. The core problems with the profiles/warnings approach is that they accumulate with an intent to be shown in the final response, but the EsqlExecutionInfo needs to be updated in real time as the search is running so that when we add `_clusters` to the async-search response of still running searches (in a later PR) it will have all the current data in the EsqlExecution info, not some possibly inaccessible rollup with the ComputeListener.

And I don't see how we can remove `runningOnRemoteCluster()` since we need to know in the refs Listener whether to populate the ComputeResponse with execution info metadata or not. The remote and local cluster refs listeners have to act differently. 

I also don't see how we can remove `isCCSListener` since the acquireCompute handler needs to know whether it is listener for a ComputeResponse with remote metadata or not, as again it needs to behave differently than a data-node handler listener. That's why I split them into separate methods in my earlier iteration of the PR, making the context or use case clear. 

Finally, another reason context is needed is because when the Status of a cluster gets set to SUCCESSFUL differs between remote and local clusters. For remote clusters, we mark it done when the remote ComputeResponse comes back. But for local, we have to wait until the coordinator is done (in case it has to run operators that aggregate/merge data from the remotes), so those get set in different places, and context/state is needed to know what code to execute in these methods that are shared for the 4 uses cases for which ComputeListener is used.

Again, I could be wrong and there's a cleaner way to do this, but I was not able to figure it out. I'd need some detailed guidance on how we solve the above problems and do local rollup accumulations like we do for profiles and warnings.

dnhatn replied:
Let's discuss this later.

dnhatn reviewed (approved):
I reviewed the production changes, and they look good to me. Thanks, Michael!

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java:
```diff
@@ -747,22 +790,33 @@ public void messageReceived(DataNodeRequest request, TransportChannel channel, T
         @Override
         public void messageReceived(ClusterComputeRequest request, TransportChannel channel, Task task) {
-            ChannelActionListener<ComputeResponse> listener = new ChannelActionListener<>(channel);
+            ChannelActionListener<ComputeResponse> lnr = new ChannelActionListener<>(channel);
```
Fixed. Thanks.

quux00 committed:
PR feedback

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 committed:
Changed status to SKIPPED when no matching index found for remote clusters

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

quux00 committed:
Merge remote-tracking branch 'elastic/main' into esql/ccs-execution-info2

elasticsearchmachine commented:
## 💚 Backport successful
| Status | Branch | Result |
|:------:|:------:|:------:|
| ✅ |  [8.x](https://github.com/elastic/elasticsearch/pull/113820)  | [<img src="https://img.shields.io/github/pulls/detail/state/elastic/elasticsearch/113820">](https://github.com/elastic/elasticsearch/pull/113820) |

quux00 commented on x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java:
```diff
@@ -519,9 +557,13 @@ private void lookupDataNodes(
             for (SearchShardsGroup group : resp.getGroups()) {
                 var shardId = group.shardId();
                 if (group.skipped()) {
+                    totalShards++;
```
Fixed.
